File: packages/@xrnjs/multi-bundle/android/src/main/java/xrn/modules/multibundle/bundle/BundleInfoManager.kt
package xrn.modules.multibundle.bundle

import android.content.Context
import androidx.annotation.RawRes
import java.io.BufferedReader
import java.io.InputStreamReader
import com.blankj.utilcode.util.GsonUtils
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import xrn.modules.multibundle.Utils

/**
 * BundleInfo 管理器
 */
object BundleInfoManager {

    const val TAG = "BundleInfoManager"

    @Volatile
    private var isInitialized = false

    private var options: BundleInfoManagerOptions? = null
    private var hook: BundleInfoHook? = null

    private var BUNDLE_INFOS: Array<BundleInfo> = arrayOf()

    private val bundleInfoMap = mutableMapOf<String, BundleInfo>();

    private var remoteBundleManager: RemoteBundleHost? = null

    fun init(options: BundleInfoManagerOptions?, hook: BundleInfoHook?) {
        Utils.assertFalse(this.isInitialized, "${TAG}.init:BundleInfoManager has initialized")
        isInitialized = true
        this.options = options
        this.hook = hook
        val bundleInfos: List<BundleInfo> = options?.bundles?.map { option ->
            val bundleType = if (option.bundleType.isNullOrBlank()) {
                BundleType.DEFAULT
            } else {
                BundleType.values().find {
                    it.name.equals(option.bundleType, ignoreCase = true)
                } ?: BundleType.DEFAULT
            }

            val bundleInfo = BundleInfo(
                option.bundleName,
                bundleType,
                option.defaultModuleName,
                option.moduleNames,
                option.codePushKey,
                option.port
            )
            bundleInfo.setHook(hook)
            if (bundleInfoMap.containsKey(bundleInfo.bundleName)) {
                throw IllegalArgumentException("${BundleInfoManager.TAG}.constructor:bundle name has set, info.bundleName=${bundleInfo.bundleName}")
            }
            bundleInfoMap[bundleInfo.bundleName] = bundleInfo
            return@map bundleInfo
        } ?: listOf()
        BUNDLE_INFOS = bundleInfos.toTypedArray()

        register(*DevBundleCache.getDevBundles().toTypedArray())
    }

    fun initWithRawFile(context: Context, @RawRes rawRes: Int, hook: BundleInfoHook? = null) {
        Utils.assertFalse(
            this.isInitialized,
            "${TAG}.initWithRawFile:BundleInfoManager has initialized"
        )
        loadOptionsFromRawFile(context, rawRes, hook)
    }

    private fun loadOptionsFromRawFile(
        context: Context,
        @RawRes rawRes: Int,
        hook: BundleInfoHook?
    ) {
        val stringBuilder = StringBuilder()
        try {
            val inputStream = context.resources.openRawResource(rawRes)
            val reader = BufferedReader(InputStreamReader(inputStream))
            var line: String?
            while ((reader.readLine().also { line = it }) != null) {
                stringBuilder.append(line)
            }
            reader.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }
        val json = stringBuilder.toString()
        val bundleInfosOption = GsonUtils.fromJson(json, BundleInfoManagerOptions::class.java)
        init(bundleInfosOption, hook)
    }

    fun prefetchRemoteBundles() {
        CoroutineScope(Dispatchers.Main).launch {
            val bundles = remoteBundleManager?.getRemoteBundles()
            bundles?.forEach { bundleInfo ->
                registerRemoteBundle(bundleInfo)
            }
        }
    }

    suspend fun getRemoteBundle(bundleName: String): BundleInfo? {
        val localBundleInfo = getBundleInfo(bundleName)
        if (localBundleInfo != null) {
            return localBundleInfo
        }

        val remoteBundleInfo = remoteBundleManager?.getRemoteBundle(bundleName)
        if (remoteBundleInfo != null) {
            registerRemoteBundle(remoteBundleInfo)
            return getBundleInfo(bundleName)
        }

        return null
    }

    fun setRemoteBundleManager(manager: RemoteBundleHost) {
        this.remoteBundleManager = manager
    }

    fun register(bundle: BundleInfo) {
        if (bundleInfoMap.containsKey(bundle.bundleName)) {
            return
        }

        bundleInfoMap[bundle.bundleName] = bundle
        BUNDLE_INFOS = BUNDLE_INFOS + bundle
    }

    fun register(vararg bundle: BundleInfo) {
        bundle.forEach {
            register(it)
        }
    }

    fun registerDevBundle(bundleName: String, port: Int) {
        val bundleInfo = BundleInfo(
            bundleName = bundleName,
            bundleType = BundleType.DEFAULT,
            defaultModuleName = "",
            moduleNames = null,
            codePushKey = null,
            port = port,
        )

        DevBundleCache.addDevBundle(bundleInfo)
        register(bundleInfo)
    }

    fun registerRemoteBundle(bundle: RemoteBundleInfo) {
        register(
            BundleInfo(
                bundleName = bundle.bundleName,
                bundleType = BundleType.DEFAULT,
                defaultModuleName = "",
                moduleNames = null,
                codePushKey = bundle.deploymentKey,
                port = 0,
                deliveryType = bundle.deliveryType
            )
        )
    }

    fun getAllBundleInfo(): List<BundleInfo> {
        Utils.assertTrue(
            this.isInitialized,
            "${TAG}.getAllBundleInfo:BundleInfoManager has not initialized"
        )
        return BUNDLE_INFOS.toList()
    }

    /**
     * 根据 bundleName 获取 BundleInfo
     * @param bundleName
     * @returns
     */
    fun getBundleInfo(bundleName: String?): BundleInfo? {
        Utils.assertTrue(
            this.isInitialized,
            "${TAG}.getBundleInfo:BundleInfoManager has not initialized"
        )
        return this.bundleInfoMap[bundleName ?: ""]
    }

    /**
     * 指定 bundle 是否已注册
     * @param bundleName
     * @returns
     */
    fun isBundleRegistered(bundleName: String): Boolean {
        Utils.assertTrue(
            this.isInitialized,
            "${TAG}.isBundleRegistered:BundleInfoManager has not initialized"
        )
        return this.getBundleInfo(bundleName) != null
    }

    /**
     * 获取 main bundle
     * @returns
     */
    fun getMainBundleInfo(): BundleInfo {
        Utils.assertTrue(
            this.isInitialized,
            "${TAG}.getMainBundleInfo:BundleInfoManager has not initialized"
        )
        val mainBundle: BundleInfo? = this.BUNDLE_INFOS.find { bundleInfo ->
            bundleInfo.isMainBundle()
        }
        return mainBundle ?: this.BUNDLE_INFOS[0]
    }

    /**
     * 根据端口获取 BundleInfo
     */
    fun findBundleInfoByPort(port: Int): BundleInfo? {
        return bundleInfoMap.values.find { it.getPort() == port }
    }
}

/**
 * BundleInfoManager 配置信息
 */
class BundleInfoManagerOptions {
    /**
     * 项目信息
     */
    var project: ProjectInfoOption? = null

    /**
     * Bundle信息
     */
    var bundles: Array<BundleInfoOption> = arrayOf()
}

/**
 * 项目信息
 */
class ProjectInfoOption {
    var name: String = ""
}

/**
 * Bundle信息
 */
class BundleInfoOption {
    /**
     * bundle name
     */
    var bundleName: String = ""

    /**
     * 是否主bundle
     */
    var bundleType: String = ""

    /**
     * 默认 moduleName，main bundle 需要配置
     */
    var defaultModuleName: String = ""

    /**
     * 所有的 bundleName
     */
    var moduleNames: Array<String> = arrayOf()

    /**
     * CodePush Key
     */
    var codePushKey: String = ""

    /**
     * 本地服务端口
     */
    var port: Int = 8081
}

File: packages/@xrnjs/multi-bundle/harmony/xrn_multi_bundle/src/main/ets/bundle/BundleInfoManager.ts
import { JSON5 } from "@wolfx/json5/src/main/ets/json5/";
import { APP_BUNDLE_BUNDLE_INFO_MANAGER } from "../Constants";
import { Context } from '@kit.AbilityKit';
import { buffer } from "@kit.ArkTS";
import { BundleInfo } from "./BundleInfo";
import { PreferencesStore } from "./PreferencesStore";

/**
 * Project 配置信息
 */
export class ProjectInfoOption {
  name: string = ""
}

/**
 * Bundle 配置信息
 */
export class BundleInfoOption {
  /**
   * bundle name
   */
  bundleName: string = ""
  /**
   * 是否主bundle
   */
  bundleType: string = ""
  /**
   * 默认 moduleName，main bundle 需要配置
   */
  defaultModuleName?: string = ""
  /**
   * 所有的 bundleName
   */
  moduleNames: string[] = []
  /**
   * CodePush Key
   */
  codePushKey: string = ""
  /**
   * 本地服务端口
   */
  port: number = 8081
}

/**
 * BundleInfoManager 配置信息
 */
export class BundleInfoManagerOptions {
  project: ProjectInfoOption | undefined
  bundles: BundleInfoOption[] = []
}

/**
 * 特殊信息Hook方法
 */
export class BundleInfoHook {
  hookCodePushKey: ((info: BundleInfo) => string) | undefined = undefined
  hookJSBundleName: ((info: BundleInfo) => string) | undefined = undefined
  hookLocalServerUrl: ((info: BundleInfo) => string) | undefined = undefined
  hookLocalServerPort: ((info: BundleInfo, initPort: number) => number) | undefined = undefined
}

const NAME_PREFERENCES = "nativeStore"
const KEY_LOCAL_BUNDLE_INFO_MAP_CACHE = "key-local-bundle-info-map-cache"

class LocalBundleInfo {
  bundleName: string;
  port: number;
}

/**
 * bundle 信息 管理类
 * 单例
 */
export class BundleInfoManager {

  private static TAG = "BundleInfoManager"

  static INSTANCE = new BundleInfoManager()

  private static isInitialized = false

  private options: BundleInfoManagerOptions | undefined
  private hook: BundleInfoHook | undefined

  private preferencesStore: PreferencesStore

  readonly BUNDLE_INFOS: BundleInfo[] = []

  readonly bundleInfoMap = new Map<string, BundleInfo>();

  private localBundleInfoMap: Map<string, LocalBundleInfo> = new Map<string, LocalBundleInfo>();


  static init(options: BundleInfoManagerOptions, preference: PreferencesStore, hook?: BundleInfoHook) {
    BundleInfoManager.assertNotInitialized('init')
    BundleInfoManager.isInitialized = true
    BundleInfoManager.INSTANCE.init(options, preference, hook)
  }

  static async initWithRawFile(context: Context, rawFilePath: string, preference: PreferencesStore, hook: BundleInfoHook | undefined) {
    BundleInfoManager.assertNotInitialized('initWithRawFile')
    BundleInfoManager.isInitialized = true
    BundleInfoManager.loadOptionsFromRawFile(context, rawFilePath, preference, hook)
  }

  private static loadOptionsFromRawFile(context: Context, rawFilePath: string, preference: PreferencesStore, hook: BundleInfoHook | undefined) {
    const  content : Uint8Array | undefined = context?.resourceManager?.getRawFileContentSync(rawFilePath)
    const result = buffer.from(content).toString('utf-8')
    const options: BundleInfoManagerOptions = JSON5.parse(result)
    BundleInfoManager.INSTANCE.init(options, preference, hook)
  }

  private constructor() {
  }

  private static assertInitialized(tag?: string) {
    if (!BundleInfoManager.isInitialized) {
      throw new Error(`${BundleInfoManager.TAG}.${tag}:BundleInfoManager has not initialized`)
    }
  }

  private static assertNotInitialized(tag?: string) {
    if (BundleInfoManager.isInitialized) {
      throw new Error(`${BundleInfoManager.TAG}.${tag}:BundleInfoManager has initialized`)
    }
  }

  private init(options: BundleInfoManagerOptions, preference: PreferencesStore, hook: BundleInfoHook | undefined) {
    this.options = options
    this.preferencesStore = preference
    this.hook = hook
    for (let index = 0; index < options.bundles.length; index++) {
      const infoOption = options.bundles[index];
      if (this.bundleInfoMap.has(infoOption.bundleName)) {
        throw new Error(`${BundleInfoManager.TAG}.constructor:bundle name has set, info.bundleName=${infoOption.bundleName}`)
      }
      const info = new BundleInfo(infoOption.bundleName, infoOption.bundleType, infoOption.defaultModuleName, infoOption.moduleNames, infoOption.codePushKey, infoOption.port)
      info.setHook(this.hook)
      this.BUNDLE_INFOS.push(info)
      this.bundleInfoMap.set(info.bundleName, info)
    }
    this.localBundleInfoMap = this.getLocalBundleInfoMapFromCache()
    this.localBundleInfoMap.forEach((value) => {
      this.registerDevBundleInfo(value.bundleName, value.port)
    })
    AppStorage.setOrCreate<BundleInfoManager>(APP_BUNDLE_BUNDLE_INFO_MANAGER, this)
  }

  getLocalBundleInfoMapFromCache(): Map<string, LocalBundleInfo> {
    const localBundleInfoMapJson = this.preferencesStore.get(KEY_LOCAL_BUNDLE_INFO_MAP_CACHE, NAME_PREFERENCES);
    const localBundleInfoMap: Map<string, LocalBundleInfo> = JSON5.parse(localBundleInfoMapJson) as Map<string, LocalBundleInfo>;
    return localBundleInfoMap;
  }

  getLocalBundleInfoMap(): Map<string, LocalBundleInfo> {
    return this.localBundleInfoMap;
  }

  saveLocalBundleList2Cache() {
    const localBundleListJson = JSON5.stringify(this.localBundleInfoMap)
    this.preferencesStore.put(KEY_LOCAL_BUNDLE_INFO_MAP_CACHE, localBundleListJson, NAME_PREFERENCES)
  }

  registerDevBundleInfo(bundleName: string, port: number) {
    if (!bundleName || !port) {
      return;
    }
    let bundleInfo = this.getBundleInfo(bundleName)
    if (bundleInfo) {
      if (bundleInfo.getDeliveryType() === BundleInfo.DELIVERY_TYPE_LOCAL) {
        // bundleName 对应的是 local bundle
        bundleInfo.setLocalServerPort(port);
        this.localBundleInfoMap[bundleName].port = port;
        this.saveLocalBundleList2Cache();
      } else {
        // bundleName 对应的是其他 bundle
        bundleInfo.setLocalServerPort(port);
        const cacheDebugJson: string = this.preferencesStore.get(`${bundleName}-debug`, NAME_PREFERENCES) as string
        const debugInfo = JSON.parse(cacheDebugJson)
        debugInfo.port = port
        const refreshDebugJson = JSON5.stringify(debugInfo)
        this.preferencesStore.put(`${bundleName}-debug`, refreshDebugJson, NAME_PREFERENCES)
      }
    } else {
      const localBundleInfo = new LocalBundleInfo()
      localBundleInfo.bundleName = bundleName
      localBundleInfo.port = port
      this.localBundleInfoMap[bundleName] = localBundleInfo
      this.saveLocalBundleList2Cache()
      const bundleInfo = this.convertLocalBundle2Bundle(localBundleInfo)
      this.registerBundleInfo(bundleInfo)
    }
  }

  private convertLocalBundle2Bundle(bundle: LocalBundleInfo): BundleInfo {
    return bundle ? new BundleInfo(bundle.bundleName, "", "", [], "", bundle.port, BundleInfo.DELIVERY_TYPE_LOCAL) : undefined
  }

  /**
   * 根据 bundleName 获取 BundleInfo
   * @param bundleName
   * @returns
   */
  getBundleInfo(bundleName: string): BundleInfo | undefined {
    BundleInfoManager.assertInitialized('getBundleInfo')
    return this.bundleInfoMap.get(bundleName)
  }

  registerBundleInfo(bundleInfo: BundleInfo): boolean {
    if (!bundleInfo || !bundleInfo.bundleName) {
      return false
    } else if (this.bundleInfoMap.has(bundleInfo.bundleName)) {
      return false
    }
    bundleInfo.setHook(this.hook)
    this.BUNDLE_INFOS.push(bundleInfo)
    this.bundleInfoMap.set(bundleInfo.bundleName, bundleInfo)
  }

  /**
   * 指定 bundle 是否已注册
   * @param bundleName
   * @returns
   */
  isBundleRegistered(bundleName: string): boolean {
    BundleInfoManager.assertInitialized('isBundleRegistered')
    return this.getBundleInfo(bundleName) ? true : false
  }

  /**
   * 获取 main bundle
   * @returns
   */
  getMainBundleInfo(): BundleInfo | null {
    BundleInfoManager.assertInitialized('getMainBundleInfo')
    const mainBundle = this.BUNDLE_INFOS.filter((value: BundleInfo, index: number, array: BundleInfo[]) => {
      return value.bundleType === BundleInfo.BUNDLE_TYPE_MAIN
    })
    return mainBundle.length > 0 ? mainBundle[0] : null
  }

  findBundleInfoByPort(port: string | number): BundleInfo | null {
    const item = this.options?.bundles.find(it => it.port.toString() === port.toString())
    if(item?.bundleName) {
      return this.getBundleInfo(item.bundleName)
    }
    return null
  }

}







